website/blog/deployers.md

```diff
@@ -0,0 +1,647 @@
+
+    // Listeners.
+    GetListenerAddress(context.Context, *protos.GetListenerAddressRequest) (*protos.GetListenerAddressReply, error)
+    ExportListener(context.Context, *protos.ExportListenerRequest) (*protos.ExportListenerReply, error)
+
+    // Telemetry.
+    HandleLogEntry(context.Context, *protos.LogEntry) error
+    HandleTraceSpans(context.Context, []trace.ReadOnlySpan) error
+}
+```
+
+Envelope initiated communication is performed by invoking a method on
+`Envelope`.
+
+```go
+// Components.
+func (e *Envelope) UpdateRoutingInfo(routing *protos.RoutingInfo) error
+func (e *Envelope) UpdateComponents(components []string) error
+
+// Telemetry.
+func (e *Envelope) GetHealth() protos.HealthStatus
+func (e *Envelope) GetMetrics() ([]*metrics.MetricSnapshot, error)
+func (e *Envelope) GetLoad() (*protos.LoadReport, error)
+func (e *Envelope) GetProfile(req *protos.GetProfileRequest) ([]byte, error)
+```
+
+## A Simple Multiprocess Deployer
+
+In this section, we implement [a fully working multiprocess
+deployer][multi_deployer]. We'll compile our deployer into an executable called
+`deploy`. We'll then be able to deploy Service Weaver binaries by running
+`./deploy <Service Weaver binary>`. To make things simple, our deployer won't
+co-locate or replicate any components. Every component will run by itself in a
+single weavelet. We begin by declaring types for the deployer and for weavelets.
+
+```go
+package main
+
+import ...
+
+// deployer is a simple multiprocess deployer that doesn't implement
+// co-location or replication. That is, every component is run in its own OS
+// process, and there is only one replica of every component.
+type deployer struct {
+    mu        sync.Mutex           // guards weavelets
+    weavelets map[string]*weavelet // weavelets, by component
+}
+
+// weavelet contains information about a weavelet. It also handles messages
+// from the weavelet by implementing the EnvelopeHandler interface.
+type weavelet struct {
+    deployer *deployer          // underlying deployer
+    envelope *envelope.Envelope // envelope to the weavelet
+    address  string             // weavelet's address
+}
+
+// Check that weavelet implements the envelope.EnvelopeHandler interface.
+var _ envelope.EnvelopeHandler = &weavelet{}
+```
+
+Next, we implement a `spawn` method that spawns a weavelet to host a component.
+
+1. To spawn the weavelet, we call the [`envelope.NewEnvelope`][NewEnvelope]
+   function. This function takes in an [`EnvelopeInfo`][EnvelopeInfo] that's
+   passed to the weavelet and an [`AppConfig`][AppConfig] that describes the
+   application. `NewEnvelope` runs the provided Service Weaver
+   binary&mdash;`flag.Arg(0)` in this case&mdash;in a subprocess. It then
+   returns an `Envelope` which communicates with the weavelet via a pair of Unix
+   pipes.
+2. We call the `UpdateComponents` method to tell the weavelet which component to
```

Comment:
Yes, I wasn't questioning the need of `UpdateComponents`.  Rather, I was wondering if we'd ever want to create a weavelet without any components.  If not, we should consider letting users specify the initial set of components to run as part of weavelet creation.  This should also make the typical `spawn` implementation simpler, since users won't have to remember calling `UpdateComponents` and we can throw an error at them if they don't specify any components in `EnvelopeInfo`.

Reply:
Ah gotcha. Let me think about this some more then. It sounds like a good idea to me.